src/main/java/org/logstash/filters/GeoIpFilter.java

```diff
@@ -0,0 +1,343 @@
+    } catch (InvalidDatabaseException e) {
+      throw new IllegalArgumentException("The database provided is invalid or corrupted.", e);
+    } catch (IOException e) {
+      throw new IllegalArgumentException("The database provided was not found in the path", e);
+    }
+    this.desiredFields = createDesiredFields(fields);
+  }
+
+  private Set<Fields> createDesiredFields(List<String> fields) {
+    Set<Fields> desiredFields = EnumSet.noneOf(Fields.class);
+    if (fields == null) {
+      switch (databaseReader.getMetadata().getDatabaseType()) {
+        case CITY_DB_TYPE:
+          desiredFields = Fields.DEFAULT_CITY_FIELDS;
+          break;
+        case COUNTRY_DB_TYPE:
+          desiredFields = Fields.DEFAULT_COUNTRY_FIELDS;
+          break;
+        case ISP_DB_TYPE:
+          desiredFields = Fields.DEFAULT_ISP_FIELDS;
+          break;
+      }
+    } else {
+      for (String fieldName : fields) {
+        desiredFields.add(Fields.parseField(fieldName));
+      }
+    }
+    return desiredFields;
+  }
+
+  public boolean handleEvent(RubyEvent rubyEvent) {
+    Event event = rubyEvent.getEvent();
+    Object input = event.getField(sourceField);
+    if (input == null) {
+      return false;
+    }
+
+    String ip;
+
+    if (input instanceof List) {
+      ip = (String) ((List) input).get(0);
+    } else if (input instanceof String) {
+      ip = (String) input;
+    } else {
+      throw new IllegalArgumentException("Expected input field value to be String or List type");
+    }
+
+    Map<String, Object> geoData = new HashMap<>();
+
+    try {
+      final InetAddress ipAddress = InetAddress.getByName(ip);
+      switch (databaseReader.getMetadata().getDatabaseType()) {
+        case CITY_DB_TYPE:
+          geoData = retrieveCityGeoData(ipAddress);
+          break;
+        case COUNTRY_DB_TYPE:
+          geoData = retrieveCountryGeoData(ipAddress);
+          break;
+        case ISP_DB_TYPE:
+          geoData = retrieveIspGeoData(ipAddress);
+          break;
+        default:
+          throw new IllegalStateException("Unsupported database type " + databaseReader.getMetadata().getDatabaseType() + "");
+      }
+    } catch (UnknownHostException e) {
+      logger.debug("IP Field contained invalid IP address or hostname. exception={}, field={}, event={}", e, sourceField, event);
+    } catch (AddressNotFoundException e) {
+      logger.debug("IP not found! exception={}, field={}, event={}", e, sourceField, event);
+    } catch (GeoIp2Exception | IOException e) {
+      logger.debug("GeoIP2 Exception. exception={}, field={}, event={}", e, sourceField, event);
+    }
+
+    return applyGeoData(geoData, event);
+  }
+
+  private boolean applyGeoData(Map<String, Object> geoData, Event event) {
+    if (geoData == null) {
+      return false;
+    }
+    // only do event.set(@target) if the lookup result is not null
+    if (event.getField(targetField) == null) {
+      event.setField(targetField, Collections.emptyMap());
```

Comment:
is this necessary? event.set("[foo][bar]", ...) should create any parent fields if it doesn't exist... I think?